### PR TITLE
Switch to a Python/YAML based build system for wheels

### PR DESCRIPTION
**Pull request by natefoo**
Many things accomplished in this PR:
1. I switched to using "official" Docker Hub images wherever possible, so that we don't have to maintain them (plus they're smaller, not sure what the magic is because generating that initial OS tarball is a dark art). Previously I was thinking it'd be better to have all the base images built in exactly the same way (and building a few base images is necessary for 32-bit support) but I don't think that's so important.
2. Debian 8+ and Ubuntu 14.04+ use the native multiarch support to avoid having to create base 32-bit images for these.
3. Added openSUSE 13, Fedora 21, and 22.
4. There are now two classes of images on which to build. The first is generic (default) which will build on Debian 6 (squeeze). The idea is that these wheels should be compatible with any other reasonable platform (sorry RHEL/CentOS 5). The second class is full, which will build on every reasonable platform I could get a docker image for. This is currently used by psycopg2 and may be used for a few others where linking to the system versions of libraries (e.g. libpq.so) would be preferred over a potentially outdated/buggy/insecure version statically linked into the wheel. **As of now I am not planning to provide a "generic" version of psycopg2**
5. Whatever images are tagged as `natefoo/base-wheel` and `natefoo/base32-wheel` will be used to build generic images.

CentOS 7 does not have a 32-bit image because it is not present in RHEL/CentOS 7. I kept 32-bit support for current versions of Debian and Ubuntu just because I figure them (or derivatives) the most likely for being installed on laptops, which is probably the only likely case where you'll find 32-bit Linux anymore. Nothing else has 32-bit support other than the generic wheels.

### CODE REVIEW

**Comment by natefoo** [comment]
The discussion on Distutils-SIG continues so this may need to change later, but this should hopefully be enough to get us out the door for now. I'd like to add openSUSE images (which I assume would build wheels compatible with SLES) since SLES is a moderately popular distro in Europe.

I'll try to get the images uploaded to docker hub at some point, but I want to take a moment to figure out why the wheel images seem to be so big (natefoo/centos-7 is only 650 MB whereas natefoo/centos7-wheel is 1.6GB).

File: wheels/image/Makefile [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+
+$(ALL_TAGS): %: %/Makefile
+	$(MAKE) -C $@ image-id
+
+$(CLEAN_TAGS):
+	$(MAKE) -C $(basename $@) clean
+
+debian6/Makefile: image-Makefile
+	mkdir -p $(@D)
+	sed -e 's#REPLACE_TAG#natefoo/$(@D)#' \
```
Comment (hexylena) [inline wheels/image/Makefile]: This seems a bit off to me, but I don't have a better solution. I tried to write a response using makefile variables (e.g. `make -f Makefile.template DIST=debian BITS=64 ...`) but that wasn't any better than this.

File: wheels/image/image-Makefile [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	echo '*.dest' >$@
+	echo 'Makefile' >>$@
+
+Dockerfile: ../Dockerfile-$(BITS).in
+	sed -e 's#IMAGE_TAR#$(DIST)#' ../Dockerfile-$(BITS).in >$@
+
+$(DIST).tar: $(DIST).dest
+	sudo tar -C $(DIST).dest -cf $(DIST).tar .
+
+debian.dest:
```
Comment (hexylena) [inline wheels/image/image-Makefile:22]: you could probably get away with a single dest and an if/else case on dist

File: wheels/image/Makefile (hunk 1 above)
Comment (natefoo) [inline wheels/image/Makefile]: Funny, this is a replacement of what I'd originally wrote, which used Makefile variables...

I was originally trying to have make generate the Makefiles, but that proved to be pretty difficult to get make to cooperate with. But I am no make expert.

File: wheels/image/image-Makefile (hunk 2 above)
Comment (natefoo) [inline wheels/image/image-Makefile:22]: True, although are 2 rules worse than an if/else? Rules are probably a bit more makeish.

Comment (hexylena) [inline wheels/image/image-Makefile:22]: yeah, fair enough. (and I was counting 4, between deb/centos)

Comment (natefoo) [inline wheels/image/image-Makefile:22]: Your're right, all considered it's 4.

**Comment by natefoo** [comment]
Now with 100% less WIP tag!

**Comment by hexylena** [comment]
LGTM/+1, but untested. Glad to see wheels are finally happening!

**Comment by natefoo** [comment]
@erasche The images are currently uploading to Docker Hub in case you actually wanted to test the wheel building. If you wanted to test the image building, glwt. If you didn't want to test anything I'll just click the 'ole merge button there after I've done a few more wheel builds of my own.

**Comment by hexylena** [comment]
@natefoo go for it

**Comment by bgruening** [comment]
Awesome! Great work @natefoo and @erasche!

**Pull request merged**